src/Stack/Upload.hs

```diff
@@ -136,34 +166,45 @@ credsFile config = do
     createDirectoryIfMissing True dir
     return $ dir </> "credentials.json"
 
-applyCreds :: HackageCreds -> Request -> IO Request
+addAPIKey :: HackageKey -> Request -> Request
+addAPIKey (HackageKey key) req =
+  setRequestHeader "Authorization" [fromString $ "X-ApiKey" ++ " " ++ T.unpack key] req
+
+applyAuth :: HasLogFunc m => HackageAuth -> Request -> RIO m Request
+applyAuth haAuth req0 = do
+    case haAuth of
+        HAKey key -> return (addAPIKey key req0)
+        HACreds creds -> applyCreds creds req0
+
+applyCreds :: HasLogFunc m => HackageCreds -> Request -> RIO m Request
 applyCreds creds req0 = do
-  manager <- getGlobalManager
-  ereq <- applyDigestAuth
+  manager <- liftIO $ getGlobalManager
+  ereq <- liftIO $ applyDigestAuth
     (encodeUtf8 $ hcUsername creds)
     (encodeUtf8 $ hcPassword creds)
     req0
     manager
   case ereq of
       Left e -> do
-          putStrLn "WARNING: No HTTP digest prompt found, this will probably fail"
+          logWarn "WARNING: No HTTP digest prompt found, this will probably fail"
           case fromException e of
-              Just e' -> putStrLn $ displayDigestAuthException e'
-              Nothing -> print e
+              Just e' -> logWarn $ fromString (displayDigestAuthException e')
+              Nothing -> logWarn $ fromString $ displayException e
```

Comment:
These two lines are so close that I wouldn't use different application styles between them. Can you make them the same? I'm partial to `$`, so that is my suggestion.
```suggestion
              Just e' -> logWarn $ fromString $ displayDigestAuthException e'
              Nothing -> logWarn $ fromString $ displayException e
```

Reply:
I didn't think of this before, but this kind of change really helps quickly spot the *true* difference between those lines!